scripts/schemagen.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		log.Fatal(fmt.Errorf("Error creating service: %v", err))
+	}
+
+	resp, err := discoveryService.Apis.GetRest(*api, *version).Fields("schemas").Do()
+	if err != nil {
+		log.Fatal(fmt.Errorf("Error reading API: %v", err))
+	}
+
+	fileName := fmt.Sprintf("gen_resource_%s_%s.go", *api, underscore(*resource))
+	f, err := os.Create(fileName)
+	defer f.Close()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fields := generateFields(resp.Schemas, *resource)
+
+	buf := &bytes.Buffer{}
+	err = googleTemplate.Execute(buf, struct {
+		TypeName string
+		Fields   map[string]string
+	}{
+		// Capitalize the first letter of the api name, then concatenate the resource name onto it.
+		// e.g. compute, instance -> ComputeInstance
+		TypeName: strings.ToUpper((*api)[0:1]) + (*api)[1:] + *resource,
+		Fields:   fields,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fmtd, err := format.Source(buf.Bytes())
+	if err != nil {
+		log.Printf("Formatting error: %s", err)
+	}
+
+	if _, err := f.Write(fmtd); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func generateFields(jsonSchemas map[string]discovery.JsonSchema, property string) map[string]string {
+	fields := make(map[string]string, 0)
+
+	for k, v := range jsonSchemas[property].Properties {
+		content, err := generateField(jsonSchemas, k, v, false)
+		if err != nil {
+			log.Printf("ERROR: %s", err)
+		} else {
+			fields[underscore(k)] = content
+		}
+	}
+
+	return fields
+}
+
+func generateField(jsonSchemas map[string]discovery.JsonSchema, field string, v discovery.JsonSchema, isNested bool) (string, error) {
+	s := &schema.Schema{
+		Description: v.Description,
+	}
+	if field != "" {
+		setProperties(v, s)
+	}
+
+	// JSON field types: https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.1
+	switch v.Type {
+	case "integer":
+		s.Type = schema.TypeInt
+	case "number":
+		s.Type = schema.TypeFloat
+	case "string":
+		s.Type = schema.TypeString
+	case "boolean":
+		s.Type = schema.TypeBool
+	case "array":
+		s.Type = schema.TypeList
+		elem, err := generateField(jsonSchemas, "", *v.Items, true)
+		if err != nil {
+			return "", fmt.Errorf("Unable to generate Elem for %q: %s", field, err)
+		}
+		s.Elem = elem
+	case "object":
+		s.Type = schema.TypeMap
+	case "":
+		s.Type = schema.TypeList
+		s.MaxItems = 1
+
+		elem := "&schema.Resource{\nSchema: map[string]*schema.Schema{\n"
+		m := generateFields(jsonSchemas, v.Ref)
+		fieldNames := []string{}
+		for k, _ := range m {
+			fieldNames = append(fieldNames, k)
+		}
+		sort.Strings(fieldNames)
+		for _, k := range fieldNames {
+			elem += fmt.Sprintf("%q: %s,\n", k, m[k])
+		}
+		elem += "},\n}"
+
+		if isNested {
+			return elem, nil
+		}
+		s.Elem = elem
+	default:
+		return "", fmt.Errorf("Unable to process: %s %s", field, v.Type)
+	}
+
+	return schemaCode(s, isNested)
+}
+
+func setProperties(v discovery.JsonSchema, s *schema.Schema) {
+	if v.ReadOnly || strings.HasPrefix(v.Description, "Output-only") {
```

Comment:
I added `[Output Only]` as an option, but I'm not going to go through the docs now and check for all possibilities (especially since many APIs don't even document required/optional/output only)